README.md
# Transmit

Plug for handling the creation of presigned urls for direct client-side uploading

## Installation

If [available in Hex](https://hex.pm/docs/publish), the package can be installed
by adding `transmit` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:transmit, "~> 0.3.0"}
  ]
end
```

Documentation can be generated with [ExDoc](https://github.com/elixir-lang/ex_doc)
and published on [HexDocs](https://hexdocs.pm). Once published, the docs can
be found at [https://hexdocs.pm/transmit](https://hexdocs.pm/transmit).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [0.3.1] - 2020-07-07

### Fixed

- Presigned URLs using virtual host option (the bucket name as part of the hostname)

## [0.3.0] - 2020-07-07

### Fixed

- Using the preferred URL structure for S3 URLs
- Make sure to use the path parameter in URL creation

## [0.2.0] - 2019-05-02

### Added

- Transmit JavaScript module

## [0.1.1] - 2019-04-10

### Fixed

- Documentation for forwarding in a Phoenix app

## [0.1.0] - 2019-04-09

### Added

- Transmit Plug
- Transmit.Signer behaviour
- Transmit.S3Signer for creating presigned URLs with S3


test/transmit_test.exs
defmodule TransmitTest do
  use ExUnit.Case
  doctest Transmit

  use ExUnit.Case, async: true
  use Plug.Test

  @opts [signer: Transmit.MockSigner]

  defmodule TestRouter do
    use Plug.Router

    plug(:match)
    plug(:dispatch)

    forward("/uploads", to: Transmit, signer: Transmit.MockSigner)
  end

  test "init" do
    assert [] = Transmit.init([])
  end

  test "without file_name param" do
    conn = Transmit.call(conn(:get, "/"), @opts)
    assert conn.status == 422
    assert conn.resp_body =~ "errors"
  end

  test "with file_name param" do
    conn = Transmit.call(conn(:get, "/?file_name=zebra.jpg"), @opts)
    assert conn.status == 200
    assert conn.resp_body =~ "data"
  end

  test "with bad file_name param" do
    conn = Transmit.call(conn(:get, "/?file_name=error.jpg"), @opts)
    assert conn.status == 422
    assert conn.resp_body =~ "errors"
  end

  test "in a router" do
    conn = TestRouter.call(conn(:get, "/uploads?file_name=zebra.jpg"), @opts)
    assert conn.status == 200
    assert conn.resp_body =~ "data"
  end
end


test/support/mock_signer.ex
defmodule Transmit.MockSigner do
  @moduledoc false
  @behaviour Transmit.Signer

  def get_presigned_url("error" <> _, _) do
    {:error, "invalid file"}
  end

  def get_presigned_url(new_file_name, _) do
    {:ok,
     %{
       presigned_url: "https://presigned.com",
       file_url: "https://s3.amazonaws.com/test/uploads/#{new_file_name}"
     }}
  end
end


mix.exs
defmodule Transmit.MixProject do
  use Mix.Project

  def project do
    [
      app: :transmit,
      version: "0.3.1",
      elixir: "~> 1.8",
      start_permanent: Mix.env() == :prod,
      deps: deps(),
      description: description(),
      elixirc_paths: elixirc_paths(Mix.env()),
      package: package(),

      # Docs
      name: "Transmit",
      source_url: "https://github.com/revelrylabs/transmit",
      homepage_url: "https://github.com/revelrylabs/transmit",
      # The main page in the docs
      docs: [main: "Transmit", extras: ["README.md"]]
    ]
  end

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:logger]
    ]
  end

  defp elixirc_paths(:test), do: ["lib", "test/support"]
  defp elixirc_paths(_), do: ["lib"]

  # Run "mix help deps" to learn about dependencies.
  defp deps do
    [
      {:plug, "~> 1.8"},
      {:uuid, "~> 1.1"},
      {:ex_aws, "~> 2.0", optional: true},
      {:ex_aws_s3, "~> 2.0", optional: true},
      {:jason, "~> 1.0", optional: true},
      {:ex_doc, "~> 0.20", only: :dev}
    ]
  end

  defp description do
    """
    Plug for handling the creation of presigned urls for direct client-side uploading
    """
  end

  defp package do
    [
      files: ["lib", "priv", "mix.exs", "README.md", "LICENSE", "CHANGELOG.md", "package.json"],
      maintainers: ["Revelry Labs"],
      licenses: ["MIT"],
      links: %{
        "GitHub" => "https://github.com/revelrylabs/transmit"
      },
      build_tools: ["mix"]
    ]
  end
end


config/config.exs
# This file is responsible for configuring your application
# and its dependencies with the aid of the Mix.Config module.
use Mix.Config

# This configuration is loaded before any dependency and is restricted
# to this project. If another project depends on this project, this
# file won't be loaded nor affect the parent project. For this reason,
# if you want to provide default values for your application for
# third-party users, it should be done in your "mix.exs" file.

# You can configure your application as:
#
#     config :transmit, key: :value
#
# and access this configuration in your application as:
#
#     Application.get_env(:transmit, :key)
#
# You can also configure a third-party app:
#
#     config :logger, level: :info
#

# It is also possible to import configuration files, relative to this
# directory. For example, you can emulate configuration per environment
# by uncommenting the line below and defining dev.exs, test.exs and such.
# Configuration from the imported file will override the ones defined
# here (which is why it is important to import them last).
#
#     import_config "#{Mix.env()}.exs"


lib/transmit/s3.ex
defmodule Transmit.S3Signer do
  @moduledoc """
  Transmit.Signer for AWS S3 Presigned URLs

  Make sure to install ex_aws and ex_aws_s3 as dependencies in your project:

  ```elixir
  def deps do
  [
    {:ex_aws, "~> 2.0"},
    {:ex_aws_s3, "~> 2.0"},
  ]
  end
  ```

  Also make sure to follow setup instructions: [https://github.com/ex-aws/ex_aws#aws-key-configuration](https://github.com/ex-aws/ex_aws#aws-key-configuration)

  ## Options
  accepts the following options:
    * `:bucket` - The S3 bucket to use
    * `:path` - the path in the bucket. (defaults to `uploads`)

  """
  @behaviour Transmit.Signer

  def get_presigned_url(new_file_name, opts) do
    bucket = Keyword.fetch!(opts, :bucket)
    path = Keyword.get(opts, :path, "uploads")

    case ExAws.S3.presigned_url(
           ExAws.Config.new(:s3),
           :put,
           bucket,
           Path.join(path, new_file_name),
           virtual_host: true
         ) do
      {:ok, presigned_url} ->
        {:ok,
         %{
           presigned_url: presigned_url,
           file_url: "https://#{bucket}.s3.amazonaws.com/#{path}/#{new_file_name}"
         }}

      {:error, _} = error ->
        error
    end
  end
end


priv/static/transmit.js
function makeS3Request(data, file) {
  return new Promise((resolve, reject) => {
    const xhr = new XMLHttpRequest()

    xhr.onreadystatechange = () => {
      if (xhr.readyState === 4) {
        if (xhr.status === 200) {
          resolve(data.url)
        } else {
          reject({
            status: xhr.status,
            statusText: xhr.statusText,
          })
        }
      }
    }

    xhr.open('PUT', data.signed_request)
    xhr.send(file)
  })
}

/**
 *
 * @param {File} file - file to upload
 * @param {string} signerPath - The path to the signer endpoint. ex: /uploads/sign
 * @returns {string} The url to the file in S3
 * @example
 * import Transmit from 'transmit'
 *
 * const imageUploadElement = document.getElementById('avatar_uploader')
 * const urlElement = document.getElementById('avatar_url')
 * // Set up so that we upload to S3 when a file is given to the imageUploadElement
 * imageUploadElement.onchange = async () => {
 *   const file = imageUploadElement.files[0]
 *   if (file === null) {
 *     alert('No file selected.')
 *   } else {
 *      const url = await Transmit.uploadFileToS3(
 *       file,
 *       '/uploads/sign'
 *     )
 *     //saved to avatar_url element
 *     urlElement.value = url
 *   }
 * }
 *
 */
async function uploadFileToS3(file, signerPath) {
  const response = await fetch(
    `${signerPath}?file_name=${file.name}&file_type=${file.type}`
  )
  const {data} = await response.json()
  const url = await makeS3Request(data, file)

  return url
}

export default {
  uploadFileToS3,
}


lib/transmit.ex
defmodule Transmit do
  @moduledoc """
  Plug for handling the creation of presigned urls for direct client-side uploading

  ## Setup

  In your router, add the following:

  ```elixir
  defmodule MyAppWeb.Router do
    use MyAppWeb, :router

    ...

    forward("/signer", Transmit, signer: Transmit.S3Signer, bucket: "images", path: "uploads")

    ...
  ```

  This makes it so that requests to `/signer` will be routed to the Transmit plug.
  The `signer` that will be used is the `Transmit.S3Signer` for creating S3 presigned URLs.
  The `bucket` and `path` options are specific to `Transmit.S3Signer` which will use them to
  create URLs that point to the `images` bucket under the path `uploads`

  For more setup information for S3 signing, make sure to check the `Transmit.S3Signer` documentation


  Different signers can be used by implementing the `Transmit.Signer` behaviour

  ## JavaScript setup

  - Add `"transmit": "file:../deps/transmit"` to your `dependencies` in your `package.json`
  - run `npm install`

  ## Usage

  ```javascript
  import Transmit from 'transmit'

  const imageUploadElement = document.getElementById('avatar_uploader')
  const urlElement = document.getElementById('avatar_url')
  // Set up so that we upload to S3 when a file is given to the imageUploadElement
  imageUploadElement.onchange = async () => {
    const file = imageUploadElement.files[0]
    if (file === null) {
      alert('No file selected.')
    } else {
       const url = await Transmit.uploadFileToS3(
        file,
        '/uploads/sign'
      )
      //saved to avatar_url element
      urlElement.value = url
    }
  }
  ```

  """

  import Plug.Conn

  def init(opts) do
    opts
  end

  def call(conn, opts) do
    conn
    |> fetch_query_params()
    |> handle(opts)
  end

  defp handle(%Plug.Conn{query_params: %{"file_name" => file_name}} = conn, opts) do
    signer = Keyword.get(opts, :signer, Transmit.S3Signer)

    uri = URI.parse(file_name)
    uuid = UUID.uuid4()

    extension =
      uri.path
      |> Path.extname()
      |> String.downcase()

    basename =
      uri.path
      |> Path.basename(extension)
      |> String.downcase()

    new_file_name = "#{basename}-#{uuid}#{extension}"
    result = signer.get_presigned_url(new_file_name, opts)

    case result do
      {:ok, %{presigned_url: signed_url, file_url: url}} ->
        response = %{
          data: %{
            signed_request: signed_url,
            file_name: new_file_name,
            file_type: MIME.type(String.replace_leading(extension, ".", "")),
            url: url
          }
        }

        json(conn, opts, 200, response)

      {:error, _} ->
        json(conn, opts, 422, %{
          errors: [
            %{
              title: "S3 Error",
              detail: "Unable to create presigned url."
            }
          ]
        })
    end
  end

  defp handle(conn, opts) do
    json(conn, opts, 422, %{
      errors: [
        %{
          title: "Missing query parameter",
          detail: "The request does not contain a 'file_name' query parameter"
        }
      ]
    })
  end

  defp json(conn, opts, status, map) do
    conn
    |> put_resp_content_type("application/json")
    |> send_resp(
      status,
      json_library(opts).encode!(map)
    )
  end

  defp json_library(opts) do
    Keyword.get(opts, :json_library, Jason)
  end
end
